File: Leetcode_Code/93. Restore IP Addresses/93_dfs.py
class Solution(object):
    def __init__(self):
        self.result = []
        self.l = []

    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        if len(s) >= 4:
            self.restoreIpAddresses_dfs(s, 0, 0)
        return self.result

    def restoreIpAddresses_dfs(self, s, start, num):
        if start >= len(s) and len(self.l) == 4:
            sb = ""
            for i in range(len(self.l)):
                sb += str(self.l[i])
                if i < len(self.l) - 1:
                    sb += '.'
            self.result.append(sb)
            return
        if 4 - num > len(s) - start or (4 - num) * 3 < len(s) - start:
            return
        for i in range(1, 4):
            if start <= len(s) - i:
                temp = int(s[start:start + i])
                if i == 1:
                    testNum = 0
                else:
                    testNum = 10 ** (i - 1)
                if testNum <= temp < 256:
                    self.l.append(temp)
                    self.restoreIpAddresses_dfs(s, start + i, num + 1)
                    self.l.pop()

File: Leetcode_Code/93. Restore IP Addresses/93_dfs.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        if (s.length() >= 4)
            restoreIpAddresses_dfs(s, 0, 0);
        return result;
    }

private:
    vector<string> result;
    vector<int> l;

    void restoreIpAddresses_dfs(string s, int start, int num) {
        if (start >= s.length() && l.size() == 4) {
            string sb = "";
            for (int i = 0; i < l.size(); ++i) {
                sb += to_string(l[i]);
                if (i < l.size() - 1)
                    sb += ".";
            }
            result.push_back(sb);
            return;
        }
        if (4 - num > s.length() - start || (4 - num) * 3 < s.length() - start)
            return;
        int temp;
        for (int i = 1; i <= 3; ++i) {
            if (start <= s.length() - i) {
                temp = stoi(s.substr((unsigned long) start, i));
                int testNum = (i == 1 ? 0 : (int) powf(10, i - 1));
                if (temp >= testNum && temp < 256) {
                    l.push_back(temp);
                    restoreIpAddresses_dfs(s, start + i, num + 1);
                    l.pop_back();
                }
            }
        }
    }
};
